### PR TITLE
Add integration test for FILE_MOUNTED_CERTS

### PR DESCRIPTION
**Pull request by nikolay-pshenichny**
This PR adds an integration test to verify that client&server can talk with each other via mTLS configured with FILE_MOUNTED_CERTS.

[ ] Configuration Infrastructure
[ ] Docs
[ ] Installation
[ ] Networking
[ ] Performance and Scalability
[ ] Policies and Telemetry
[ ] Security
[x] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

Pull Request Attributes

Please check any characteristics that apply to this pull request. 

[ ] Does not have any changes that may affect Istio users.

### CODE REVIEW

File: pkg/test/framework/components/istio/operator.go [hunk 1]
```diff
@@ -480,7 +480,7 @@ func installControlPlaneCluster(i *operatorComponent, cfg Config, cluster resour
 		}
 	}
 
-	if i.environment.IsConfigCluster(cluster) {
+	if i.environment.IsMulticluster() && i.environment.IsConfigCluster(cluster) {
```
Comment (nikolay-pshenichny) [inline pkg/test/framework/components/istio/operator.go:483]: Without this change, "eastwest-gateway" seems to be installed even for the single cluster scenarios.
cc: @ramaraochavali

Comment (howardjohn) [reply pkg/test/framework/components/istio/operator.go:483]: Its used for VMs as well

File: tests/integration/security/file_mounted_certs/main_test.go [hunk 2]
```diff
@@ -0,0 +1,206 @@
...
+                  defaultMode: 420
+            - path: spec.template.spec.containers[name:discovery].volumeMounts[7]
+              value: |-
+                name: server-certs
+                mountPath: /server-certs
+
+meshConfig:
+  defaultConfig:
+    controlPlaneAuthPolicy: "NONE"
+    discoveryAddress: istiod.istio-system:15010
```
Comment (howardjohn) [inline tests/integration/security/file_mounted_certs/main_test.go:127]: why are we using plaintext?

File: tests/integration/security/file_mounted_certs/main_test.go [hunk 3]
```diff
@@ -0,0 +1,206 @@
...
+    pilotCertProvider: "mycopki"
+    logging:
+      level: "default:debug"
+    proxy:
+      logLevel: debug
+      componentLogLevel: "misc:debug"
+  pilot:
+    env:
+      PILOT_SKIP_VALIDATE_TRUST_DOMAIN: "true"
+      PILOT_SIDECAR_USE_REMOTE_ADDRESS: "true"
```
Comment (howardjohn) [inline tests/integration/security/file_mounted_certs/main_test.go:141]: this doesn't seem related?

File: tests/integration/security/file_mounted_certs/main_test.go [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+  global:
+    pilotCertProvider: "mycopki"
+    logging:
+      level: "default:debug"
+    proxy:
+      logLevel: debug
+      componentLogLevel: "misc:debug"
+  pilot:
+    env:
+      PILOT_SKIP_VALIDATE_TRUST_DOMAIN: "true"
```
Comment (howardjohn) [inline tests/integration/security/file_mounted_certs/main_test.go:140]: why are we skipping?

File: tests/integration/security/file_mounted_certs/main_test.go (hunk 2 above)
Comment (nikolay-pshenichny) [reply tests/integration/security/file_mounted_certs/main_test.go:127]: @howardjohn , the intent was to test mtls between the client and server and simplify everything else by using plaintext with control plane.
Let me change this.

File: tests/integration/security/file_mounted_certs/main_test.go (hunk 4 above)
Comment (nikolay-pshenichny) [reply tests/integration/security/file_mounted_certs/main_test.go:140]: The test certificates that are used as server/client certs don't have `spiffe://cluster.local` in SANs; and if Trust Domain validation is not skipped, test Echo server will refuse connections due to:
```
            "combined_validation_context": {
             "default_validation_context": {
              "match_subject_alt_names": [
               {
                "prefix": "spiffe://cluster.local/"
               }
              ]
             },

```

I looked into this a bit, and It seems that i can try using the `default` certs as "client certs". Those have the Spiffe URI in SANs. Let me make that change.

File: tests/integration/security/sds_ingress/util/util.go [hunk 5]
```diff
@@ -19,8 +19,8 @@ import (
 	"bytes"
 	"context"
 	"fmt"
-	"strconv"
 	"strings"
+	"testing"
```
Comment (nikolay-pshenichny) [inline tests/integration/security/sds_ingress/util/util.go:23]: This file should not have any changes. will revert this.

File: tests/integration/security/file_mounted_certs/main_test.go (hunk 4 above)
Comment (nikolay-pshenichny) [reply tests/integration/security/file_mounted_certs/main_test.go]: fixed

File: tests/integration/security/file_mounted_certs/main_test.go (hunk 3 above)
Comment (nikolay-pshenichny) [reply tests/integration/security/file_mounted_certs/main_test.go]: removed

File: tests/integration/security/file_mounted_certs/main_test.go (hunk 2 above)
Comment (nikolay-pshenichny) [reply tests/integration/security/file_mounted_certs/main_test.go]: updated the PR to use mTLS with istiod

File: pkg/test/framework/components/istio/operator.go (hunk 1 above)
Comment (nikolay-pshenichny) [reply pkg/test/framework/components/istio/operator.go:474]: @howardjohn , sorry, I am not completely following.
Is the change here ok to keep, or it will break VMs?

Comment (nikolay-pshenichny) [reply pkg/test/framework/components/istio/operator.go]: Found an answer in the results of the integration tests.
Reverting this change fixes the issue.

But, it seems that eastwest-gateway is even being installed for the test added in this PR. @howardjohn , do you know if there's a way to somehow disable it while initializing a NewSuite() - https://github.com/istio/istio/pull/27861/files#diff-b8125949e889b8b57cb90c1f16049b984e751c4cfac683ca1915224ca24ff65aR63

File: tests/integration/security/file_mounted_certs/main_test.go [hunk 6]
```diff
@@ -0,0 +1,204 @@
...
+	"k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+var (
+	inst istio.Instance
+)
+
+const (
+	ProxyMetadataJson = `
```
Comment (howardjohn) [inline tests/integration/security/file_mounted_certs/main_test.go:43]: I would avoid adding things other than exactly what we want to test. This makes it less confusing what the intent of the test is, and what are required and "optional" settings.

Namely, jwt policy, xds via agent, dns capture, and stats settings

File: tests/integration/security/file_mounted_certs/main_test.go [hunk 7]
```diff
@@ -0,0 +1,204 @@
...
+
+func setupConfig(_ resource.Context, cfg *istio.Config) {
+	if cfg == nil {
+		return
+	}
+
+	cfg.ControlPlaneValues = `
+components:
+  egressGateways:
+  - enabled: false
```
Comment (howardjohn) [inline tests/integration/security/file_mounted_certs/main_test.go:82]: Same comment as above, I see no reason to disable this - if anything, it gives us more coverage.

File: tests/integration/security/file_mounted_certs/main_test.go [hunk 8]
```diff
@@ -0,0 +1,204 @@
...
+    name: istio-ingressgateway
+  pilot:
+    enabled: true
+    k8s:
+      overlays:
+        - kind: Deployment
+          name: istiod
+          patches:
+            - path: spec.template.spec.containers.[name:discovery].args
```
Comment (howardjohn) [inline tests/integration/security/file_mounted_certs/main_test.go:94]: nit: does the overlay allow appending or only replacing? If appending we should prefer this

File: tests/integration/security/file_mounted_certs/main_test.go [hunk 9]
```diff
@@ -0,0 +1,204 @@
...
+                  defaultMode: 420
+            - path: spec.template.spec.containers[name:discovery].volumeMounts[7]
+              value: |-
+                name: server-certs
+                mountPath: /server-certs
+
+meshConfig:
+  defaultConfig:
+    controlPlaneAuthPolicy: "MUTUAL_TLS"
+    discoveryAddress: istiod.istio-system:15012
```
Comment (howardjohn) [inline tests/integration/security/file_mounted_certs/main_test.go:127]: this seems like it should be removed

File: tests/integration/security/file_mounted_certs/main_test.go [hunk 10]
```diff
@@ -0,0 +1,204 @@
...
+    proxy:
+      logLevel: debug
+      componentLogLevel: "misc:debug"
+  pilot:
+    env:
+      PILOT_ENABLE_PROTOCOL_SNIFFING_FOR_OUTBOUND: "true"
+      PILOT_ENABLE_PROTOCOL_SNIFFING_FOR_INBOUND: "true"
+      XDS_AUTH: "false"
+      PILOT_ENABLE_CRD_VALIDATION: "true"
+      DNS_ADDR: ""
```
Comment (howardjohn) [inline tests/integration/security/file_mounted_certs/main_test.go:144]: Remove all of this?

File: tests/integration/security/file_mounted_certs/p2p_mtls_test.go [hunk 11]
```diff
@@ -0,0 +1,410 @@
...
+metadata:
+  name: server
+  namespace: {{.AppNamespace}}
+spec:
+  hosts:
+  - server.{{.AppNamespace}}.svc.cluster.local
+  http:
+  - retries:
+      attempts: 5
+      retryOn: cancelled,connect-failure,gateway-error,refused-stream,resource-exhausted,unavailable,retriable-status-codes
```
Comment (howardjohn) [inline tests/integration/security/file_mounted_certs/p2p_mtls_test.go:140]: do we need this?

File: tests/integration/security/file_mounted_certs/p2p_mtls_test.go [hunk 12]
```diff
@@ -0,0 +1,410 @@
...
+      subjectAltNames:
+        - server.default.svc
+
+`
+	VirtualServiceConfigMutual = `
+apiVersion: networking.istio.io/v1beta1
+kind: VirtualService
+metadata:
+  name: server
+  namespace: {{.AppNamespace}}
```
Comment (howardjohn) [inline tests/integration/security/file_mounted_certs/p2p_mtls_test.go:133]: nit: you can just apply it to AppNamespace and use short names if you want to avoid templating

File: tests/integration/security/file_mounted_certs/p2p_mtls_test.go [hunk 13]
```diff
@@ -0,0 +1,410 @@
...
+kind: PeerAuthentication
+metadata:
+  name: default
+  namespace: "istio-system"
+spec:
+  mtls:
+    mode: STRICT
+`
+
+	EnvoyFilterConfig = `
```
Comment (howardjohn) [inline tests/integration/security/file_mounted_certs/p2p_mtls_test.go:159]: why is this removed?

File: tests/integration/security/file_mounted_certs/p2p_mtls_test.go [hunk 14]
```diff
@@ -0,0 +1,410 @@
...
+      operation: REMOVE
+`
+)
+
+
+func createObject(t *testing.T, ctx framework.TestContext, serviceNamespace string, yamlManifest string) bytes.Buffer {
+	var destinationRuleToParse string
+	destinationRuleToParse = DestinationRuleConfigMutual
+
+	tmpl, err := template.New("DestinationRule").Parse(destinationRuleToParse)
```
Comment (howardjohn) [inline tests/integration/security/file_mounted_certs/p2p_mtls_test.go:189]: nit: I would `tmpl.EvaluateOrFail`

File: tests/integration/security/file_mounted_certs/p2p_mtls_test.go [hunk 15]
```diff
@@ -0,0 +1,410 @@
...
+			expectedValue: 1.0,
+		},
+		{
+			metricName: "envoy_listener_http_inbound_0_0_0_0_8443_downstream_rq_completed",
+			expectedValue: 1.0,
+		},
+	}
+
+	for _, tc := range serverTestCases {
+		metric = getMetricForListener(stats[tc.metricName].Metric, "0.0.0.0_15006") // Check metrics on VirtualInbound listener
```
Comment (howardjohn) [inline tests/integration/security/file_mounted_certs/p2p_mtls_test.go:405]: This seems fairly complex. Do we get any value checking metrics here, that the above testing with setting STRICT and checking for the client cert header do not give us?

File: tests/integration/security/file_mounted_certs/p2p_mtls_test.go [hunk 16]
```diff
@@ -0,0 +1,410 @@
...
+			if (*label.Name == "listener_address" && *label.Value == listenerName) {
+				return metric
+			}
+		}
+	}
+	return nil
+}
+
+
+func verifyUsingFileBasedCerts(t *testing.T, client echo.Instance, server echo.Instance) {
```
Comment (howardjohn) [inline tests/integration/security/file_mounted_certs/p2p_mtls_test.go:302]: Won't the test fail with cert mismatch if they aren't? IMO integration tests should be testing behavior - if you want to test the generated XDS output it should be in a unit test

**Review by howardjohn — COMMENTED** [review]
cc @nmittler

File: tests/integration/security/file_mounted_certs/main_test.go (hunk 8 above)
Comment (nikolay-pshenichny) [reply tests/integration/security/file_mounted_certs/main_test.go:94]: I tried a few variations and the following seems to work:
```
            - path: spec.template.spec.containers.[name:discovery].args[1001]
              value: "--caCertFile=/server-certs/ca.pem"
            - path: spec.template.spec.containers.[name:discovery].args[1002]
              value: "--tlsCertFile=/server-certs/cert.pem"
            - path: spec.template.spec.containers.[name:discovery].args[1003]
              value: "--tlsKeyFile=server-certs/key.pem"

```

I'll update the code.

File: tests/integration/security/file_mounted_certs/main_test.go (hunk 7 above)
Comment (nikolay-pshenichny) [reply tests/integration/security/file_mounted_certs/main_test.go:82]: Will remove these settings.
Originally my thought was that If IG is not needed for the test it is safe to disable it.
thx!

File: tests/integration/security/file_mounted_certs/p2p_mtls_test.go (hunk 16 above)
Comment (nikolay-pshenichny) [reply tests/integration/security/file_mounted_certs/p2p_mtls_test.go:302]: In theory it should fail, yes.
But I keep thinking about a scenario when, for some reason, the ISTIO_META_CERT_X params were not recognized and everything is still functioning with the default setup and not using File Mounted Certs.
That's the reason I wanted to have this verification here.

File: tests/integration/security/file_mounted_certs/p2p_mtls_test.go (hunk 15 above)
Comment (nikolay-pshenichny) [reply tests/integration/security/file_mounted_certs/p2p_mtls_test.go:405]: Thinking about this more - you right. STRICT + proper xfcc validation will be enough.
I just need to provide properly tailored certs.
Right now this test is reusing DNS&Pilot certs in different combinations. Let me update this.
Thank you!

File: tests/integration/security/file_mounted_certs/p2p_mtls_test.go (hunk 13 above)
Comment (nikolay-pshenichny) [reply tests/integration/security/file_mounted_certs/p2p_mtls_test.go:159]: it's not needed, will remove this.
thank you.

File: tests/integration/security/file_mounted_certs/main_test.go (hunk 6 above)
Comment (nikolay-pshenichny) [reply tests/integration/security/file_mounted_certs/main_test.go:43]:
```
"PROXY_XDS_VIA_AGENT": "false",
```
was needed due to the fact that i reuse existing test certificates in this test and they don't have enough info.
without this config i was seeing the following error:

```
	info	Could not verify certificate: peer certificate does not contain 1 URI type SAN, detected 0
```

Going to add proper test certs and that will allow to remove most of the custom settings.

File: tests/integration/security/file_mounted_certs/main_test.go [hunk 17]
```diff
@@ -0,0 +1,177 @@
...
+  global:
+    pilotCertProvider: "mycopki"
+    logging:
+      level: "default:debug"
+    proxy:
+      logLevel: debug
+      componentLogLevel: "misc:debug"
+  pilot:
+    env:
+      XDS_AUTH: "false"
```
Comment (nikolay-pshenichny) [inline tests/integration/security/file_mounted_certs/main_test.go:118]: @howardjohn , I left `XDS_AUTH: false` here because without it server/client connections to istiod will fail with the following errors:

```
2020-10-24T00:17:22.684038Z	warn	ads	Unauthorized XDS: 10.1.0.154:43728 with identity [spiffe://cluster.local/ns/mounted-certs/sa/server server.mounted-certs.svc]: no identities ([spiffe://cluster.local/ns/mounted-certs/sa/server server.mounted-certs.svc]) matched istio-fd-sds-1-77818/default
  2020-10-24T00:17:24.051879Z	warn	ads	Unauthorized XDS: 10.1.0.159:41960 with identity [spiffe://cluster.local/ns/mounted-certs/sa/client client.mounted-certs.svc]: no identities ([spiffe://cluster.local/ns/mounted-certs/sa/client client.mounted-certs.svc]) matched istio-fd-sds-1-4523/default
```

Ie. the expected identity is not in the test certificates. Please let me know if there's a way around this.

**Comment by nikolay-pshenichny** [comment]
@howardjohn , I updated the PR, please take a look when you have a chance. Thank you.

File: tests/integration/security/file_mounted_certs/main_test.go (hunk 17 above)
Comment (howardjohn) [reply tests/integration/security/file_mounted_certs/main_test.go:175]: I think thats tricky since the namespace may not be fixed. lets keep this for now, but would be good to add a comment about why we did it

**Review by howardjohn — COMMENTED** [review]
LGTM

cc @JimmyCYJ @williamaronli @lei-tang can you review/approve?

**Review by lei-tang — APPROVED** [review]

**Comment by lei-tang** [comment]
@nikolay-pshenichny LGTM, thank you for your PR! Please fix the conflicting files error.

**Review by howardjohn — APPROVED** [review]
